logdetective/server/plot.py

```diff
@@ -0,0 +1,206 @@
+        color="skyblue",
+        label="Requests",
+    )
+    ax1.set_xlabel("Time")
+    ax1.set_ylabel("Requests", color="blue")
+    ax1.tick_params(axis="y", labelcolor="blue")
+
+    ax1.xaxis.set_major_formatter(matplotlib.dates.DateFormatter(plot_def.time_format))
+    ax1.xaxis.set_major_locator(plot_def.locator)
+
+    matplotlib.pyplot.xticks(rotation=45)
+
+    ax1.grid(True, alpha=0.3)
+
+
+def _add_cumulative_line_for_requests_count(
+    ax2: matplotlib.figure.Axes, timestamps: numpy.array, counts: numpy.array
+) -> None:
+    """Add cumulative line on secondary y-axis"""
+    cumulative = numpy.cumsum(counts)
+    ax2.plot(timestamps, cumulative, "r-", linewidth=2, label="Cumulative")
+    ax2.set_ylabel("Cumulative Requests", color="red")
+    ax2.tick_params(axis="y", labelcolor="red")
+
+
+def requests_per_time(
+    period_of_time: models.TimePeriod,
+    endpoint: EndpointType = EndpointType.ANALYZE,
+    end_time: Optional[datetime.datetime] = None,
+) -> matplotlib.figure.Figure:
+    """
+    Generate a visualization of request counts over a specified time period.
+
+    This function creates a dual-axis plot showing:
+    1. A bar chart of request counts per time interval
+    2. A line chart showing the cumulative request count
+
+    The time intervals are determined by the provided TimePeriod object, which defines
+    the granularity and formatting of the time axis.
+
+    Args:
+        period_of_time: A TimePeriod object that defines the time period and interval
+                        for the analysis (e.g., hourly, daily, weekly)
+        endpoint: One of the API endpoints
+        end_time: The end time for the analysis period. If None, defaults to the current
+                  UTC time
+
+    Returns:
+        A matplotlib Figure object containing the generated visualization
+    """
+    end_time = end_time or datetime.datetime.now(datetime.timezone.utc)
+    start_time = period_of_time.get_period_start_time(end_time)
+    plot_def = Definition(period_of_time)
+    requests_counts = AnalyzeRequestMetrics.get_requests_in_period(
+        start_time, end_time, plot_def.time_format, endpoint
+    )
+    timestamps, counts = create_time_series_arrays(
+        requests_counts, start_time, end_time, plot_def.time_delta, plot_def.time_format
+    )
+
+    fig, ax1 = matplotlib.pyplot.subplots(figsize=(12, 6))
+    _add_bar_chart_for_requests_count(ax1, plot_def, timestamps, counts)
+
+    ax2 = ax1.twinx()
+    _add_cumulative_line_for_requests_count(ax2, timestamps, counts)
+
+    matplotlib.pyplot.title(
+        f"Requests received for API {endpoint} ({start_time.strftime(plot_def.time_format)} "
+        f"to {end_time.strftime(plot_def.time_format)})"
+    )
+
+    lines1, labels1 = ax1.get_legend_handles_labels()
+    lines2, labels2 = ax2.get_legend_handles_labels()
+    ax1.legend(lines1 + lines2, labels1 + labels2, loc="center")
+
+    matplotlib.pyplot.tight_layout()
+
+    return fig
```

Comment:
Maja, this is super impressive!! I expected a "stupid" local script that we would call in some funky way. Having this integrated directly in the site is mindblowing!

Reply:
Thanks Tomas, I think this approach is quicker and more convenient but I am not sure this is what we want. @jpodivin I am mistaken or you are concerned about exposing the metrics in the server? (I got this feeling in the yesterday standup). I can move this same code in a script if you prefer.

Reply:
I'm somewhat concerned about people spamming it, thus causing issues with response times for the server as a whole. That being said, I understand that it may be a premature worry. I suppose we can merge this and later maybe split this off into it's own server or add authentication.

Reply:
For the future I believe we should send the metrics to Grafana (using the Pushgateway) or something similar. I believe these plots and their endpoints can be trow away when we will have the necessary infrastructure for using Grafana.